lib/puppet/ssl/state_machine.rb

```diff
@@ -56,32 +72,177 @@ def next_state
+        next_ctx = @ssl_provider.create_root_context(cacerts: ssl_context[:cacerts], crls: [])
+      end
+
+      NeedKey.new(@machine, next_ctx)
+    rescue Puppet::Rest::ResponseError => e
+      if e.response.code.to_i == 404
+        raise Puppet::Error.new(_('CRL is missing from the server'))
+      else
+        raise Puppet::Error.new(_('Could not download CRLs: %{message}') % { message: e.message }, e)
+      end
+    end
+  end
+
+  # Load or generate a private key. If the key exists, try to load the client cert
+  # and transition to Done. If the cert is mismatched or otherwise fails valiation,
+  # raise an error. If the key doesn't exist yet, generate one, and save it. If the
+  # cert doesn't exist yet, transition to NeedSubmitCSR.
+  #
+  class NeedKey < SSLState
+    def next_state
+      key = @cert_provider.load_private_key(Puppet[:certname])
+      if key
+        cert = @cert_provider.load_client_cert(Puppet[:certname])
+        if cert
+          next_ctx = @ssl_provider.create_context(
+            cacerts: @ssl_context.cacerts, crls: @ssl_context.crls, private_key: key, client_cert: cert
+          )
+          return Done.new(@machine, next_ctx)
+        end
+      else
+        Puppet.info _("Creating a new SSL key for %{name}") % { name: Puppet[:certname] }
+        key = OpenSSL::PKey::RSA.new(Puppet[:keylength].to_i)
+        @cert_provider.save_private_key(Puppet[:certname], key)
+      end
+
+      NeedSubmitCSR.new(@machine, @ssl_context, key)
+    end
+  end
+
+  # Base class for states with a private key.
+  #
+  class KeySSLState < SSLState
+    attr_reader :private_key
+
+    def initialize(machine, ssl_context, private_key)
+      super(machine, ssl_context)
+      @private_key = private_key
+    end
+  end
+
+  # Generate and submit a CSR using the CA cert bundle and optional CRL bundle
+  # from earlier states. If the request is submitted, proceed to NeedCert,
+  # otherwise Wait. This could be due to the server already having a CSR
+  # for this host (either the same or different CSR content), having a
+  # signed certificate, or a revoked certificate.
+  #
+  class NeedSubmitCSR < KeySSLState
+    def next_state
+      csr = @cert_provider.create_request(Puppet[:certname], @private_key)
+      Puppet::Rest::Routes.put_certificate_request(csr.to_pem, Puppet[:certname], @ssl_context)
+      @cert_provider.save_request(Puppet[:certname], csr)
+      NeedCert.new(@machine, @ssl_context, @private_key)
+    rescue Puppet::Rest::ResponseError => e
+      if e.response.code.to_i != 400
+        raise Puppet::SSL::SSLError.new(_("Failed to submit the CSR, HTTP response was %{code}") % { code: e.response.code }, e)
       end
 
-      NeedKey.new(next_ctx)
+      NeedCert.new(@machine, @ssl_context, @private_key)
     end
   end
 
-  class NeedKey < SSLState; end
+  # Attempt to load or retrieve our signed cert.
+  #
+  class NeedCert < KeySSLState
+    def next_state
+      cert = OpenSSL::X509::Certificate.new(
```

Comment:
If there was a matching private key and cert, we would have gone from NeedKey->Done. If there wasn't a cert or it mismatched, then we'll end up here, and correctly overwrite the cert.

Reply:
OK, NeedKey -> Done if the cert exists and gets verified.  If not, it goes to NeedSubmitCSR which either raises or goes to NeedCert.  NeedCert goes to Done if the server returns the cert immediately, otherwise goes to Wait.